arwen/elrondapi/elrondei.go

```diff
@@ -1776,6 +1776,11 @@ func upgradeContract(
 		value,
 	)
 	logEEI.Trace("upgradeContract", "error", err)
+
+	storage := host.Storage()
+	if storage.IsUseDifferentGasCostFlagSet() && arwen.WithFaultAndHost(host, err, runtime.ElrondAPIErrorShouldFailExecution()) {
```

Comment:
I strongly recommend integrating the following condition:
```
	if errors.Is(err, arwen.ErrNotEnoughGas) {
		runtime.SetRuntimeBreakpointValue(arwen.BreakpointOutOfGas)
		return
	}
```

See `v1_4_asyncCall()` below, which does this.